driver-core/src/main/com/mongodb/internal/connection/PowerOfTwoBufferPool.java

```diff
@@ -51,27 +110,37 @@ public PowerOfTwoBufferPool() {
      * @param highestPowerOfTwo the highest power of two buffer size that will be pooled
      */
     public PowerOfTwoBufferPool(final int highestPowerOfTwo) {
+        this(highestPowerOfTwo, DEFAULT_MAX_IDLE_TIME_MINUTES, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Construct an instance.
+     *
+     * @param highestPowerOfTwo the highest power of two buffer size that will be pooled
+     * @param maxIdleTime max idle time when pruning is enabled
+     * @param timeUnit time unit of maxIdleTime
+     */
+    public PowerOfTwoBufferPool(final int highestPowerOfTwo, final long maxIdleTime, final TimeUnit timeUnit) {
         int powerOfTwo = 1;
         for (int i = 0; i <= highestPowerOfTwo; i++) {
-            final int size = powerOfTwo;
-            powerOfTwoToPoolMap.put(i, new ConcurrentPool<>(INFINITE_SIZE,
-                                                                         new ConcurrentPool.ItemFactory<ByteBuffer>() {
-                                                                             @Override
-                                                                             public ByteBuffer create() {
-                                                                                 return createNew(size);
-                                                                             }
-
-                                                                             @Override
-                                                                             public void close(final ByteBuffer byteBuffer) {
-                                                                             }
-
-                                                                             @Override
-                                                                             public Prune shouldPrune(final ByteBuffer byteBuffer) {
-                                                                                 return Prune.STOP;
-                                                                             }
-                                                                         }));
+            int size = powerOfTwo;
+            powerOfTwoToPoolMap.put(i, new ConcurrentPool<>(INFINITE_SIZE, new ItemFactory(size)));
             powerOfTwo = powerOfTwo << 1;
         }
+        maxIdleTimeNanos = timeUnit.toNanos(maxIdleTime);
+        pruner = Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory("BufferPoolPruner"));
+    }
+
+    /**
+     * Call this method at most once to enable a background thread that prunes idle buffers from the pool
+     */
+    public void enablePruning() {
+        pruner.scheduleAtFixedRate(this::prune, maxIdleTimeNanos, maxIdleTimeNanos / 2, TimeUnit.NANOSECONDS);
```

Comment:
Using `maxIdleTimeNanos / 2` instead of `maxIdleTimeNanos` as the period reduces the practically achievable "max idle time" from roughly `2 * maxIdleTimeNanos` to roughly `1.5 * maxIdleTimeNanos`. Pointing this out just to make sure we have the same understanding: despite `maxIdleTimeNanos` is set to 1 minute, idle buffers may stay in the pool for up to 1.5 minutes even without any unexpected freezes caused by the GC or the OS scheduler.

Reply:
Yes, that sounds right.  I'm ok with it since 1 minute is not a publicly settable value, so anything we do here is arbitrary.  We could always run the thread more often.